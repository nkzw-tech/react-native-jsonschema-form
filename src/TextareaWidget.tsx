import {
  ariaDescribedByIds,
  FormContextType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
} from '@rjsf/utils';
import { View } from 'react-native';
import { Textarea } from './components/ui/textarea.tsx';

type CustomWidgetProps<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any,
> = WidgetProps<T, S, F> & {
  options: any;
};

/** The `TextareaWidget` is a widget for rendering input fields as textarea.
 *
 * @param props - The `WidgetProps` for this component
 */
export default function TextareaWidget<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any,
>({
  autofocus,
  id,
  onChange,
  options,
  placeholder,
  readonly,
  value,
}: CustomWidgetProps<T, S, F>) {
  const _onChange = (value: string) =>
    onChange(value === '' ? options.emptyValue : value);

  return (
    <View className="flex-row">
      <Textarea
        aria-describedby={ariaDescribedByIds<T>(id)}
        autoFocus={autofocus}
        id={id}
        numberOfLines={options.rows || 5}
        onChangeText={_onChange}
        placeholder={placeholder}
        readOnly={readonly}
        value={value ?? ''}
      />
    </View>
  );
}
