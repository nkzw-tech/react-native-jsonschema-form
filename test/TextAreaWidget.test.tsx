import renderer from 'react-test-renderer';
import TextareaWidget from '../src/TextareaWidget.tsx';
import { makeWidgetMockProps } from './helpers/createMocks.ts';

describe('TextareaWidget', () => {
  test('simple without errors', () => {
    const tree = renderer
      .create(<TextareaWidget {...makeWidgetMockProps({})} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('simple with errors', () => {
    const tree = renderer
      .create(
        <TextareaWidget
          {...makeWidgetMockProps({ rawErrors: ['Invalid 1', 'Invalid 2'] })}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('simple without required', () => {
    const tree = renderer
      .create(<TextareaWidget {...makeWidgetMockProps({ required: false })} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
