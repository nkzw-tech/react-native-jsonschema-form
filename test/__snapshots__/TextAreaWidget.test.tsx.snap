// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TextareaWidget simple with errors 1`] = `
<View
  className="flex-row"
>
  <TextInput
    aria-describedby="_id__error _id__description _id__help"
    autoFocus={true}
    className="w-full flex-row rounded-md border border-input bg-transparent p-3 text-sm placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
    id="_id"
    numberOfLines={5}
    onChangeText={[Function]}
    placeholder=""
    readOnly={true}
    value="value"
  />
</View>
`;

exports[`TextareaWidget simple without errors 1`] = `
<View
  className="flex-row"
>
  <TextInput
    aria-describedby="_id__error _id__description _id__help"
    autoFocus={true}
    className="w-full flex-row rounded-md border border-input bg-transparent p-3 text-sm placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
    id="_id"
    numberOfLines={5}
    onChangeText={[Function]}
    placeholder=""
    readOnly={true}
    value="value"
  />
</View>
`;

exports[`TextareaWidget simple without required 1`] = `
<View
  className="flex-row"
>
  <TextInput
    aria-describedby="_id__error _id__description _id__help"
    autoFocus={true}
    className="w-full flex-row rounded-md border border-input bg-transparent p-3 text-sm placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
    id="_id"
    numberOfLines={5}
    onChangeText={[Function]}
    placeholder=""
    readOnly={true}
    value="value"
  />
</View>
`;
